File: ios-app/BayitPlusApp/Models/AudiobookModels.swift
import Foundation

// MARK: - Audiobooks

/// An audiobook with metadata and chapter information.
struct Audiobook: Decodable, Sendable, Identifiable {
    let id: String
    let title: String?
    let author: String?
    let narrator: String?
    let description: String?
    let thumbnail: String?
    let backdrop: String?
    let duration: String?
    let chapters: [AudiobookChapter]?
    let totalChapters: Int?
    let genreIds: [String]?
    let audioQuality: String?
    let requiresSubscription: String?
    let contentFormat: String?
    let streamUrl: String?
    let streamType: String?
    let viewCount: Int?
    let avgRating: Double?
    let isFeatured: Bool?
    let createdAt: String?
    let updatedAt: String?
}

/// A chapter within an audiobook.
struct AudiobookChapter: Decodable, Sendable, Identifiable {
    let id: String?
    let title: String?
    let chapterNumber: Int?
    let duration: String?
    let progress: Double?
    let thumbnail: String?
    let streamUrl: String?
    let streamType: String?
    let startTime: Double?
    let endTime: Double?

    var stableId: String { id ?? "\(chapterNumber ?? 0)" }
}

/// Paginated response from GET /api/v1/audiobooks
struct AudiobookListResponse: Decodable, Sendable {
    let items: [Audiobook]?
    let total: Int?
    let page: Int?
    let pageSize: Int?
    let totalPages: Int?
}

File: android-app/core/core-model/src/main/java/tv/bayit/plus/core/model/AudiobookModels.kt
package tv.bayit.plus.core.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/** An audiobook with metadata and chapter information. */
@Serializable
data class Audiobook(
    val id: String,
    val title: String? = null,
    val author: String? = null,
    val narrator: String? = null,
    val description: String? = null,
    val thumbnail: String? = null,
    val backdrop: String? = null,
    val duration: String? = null,
    val chapters: List<AudiobookChapter>? = null,
    @SerialName("genre_ids") val genreIds: List<String>? = null,
    @SerialName("audio_quality") val audioQuality: String? = null,
    @SerialName("requires_subscription") val requiresSubscription: String? = null,
    @SerialName("content_format") val contentFormat: String? = null,
    @SerialName("view_count") val viewCount: Int? = null,
    @SerialName("avg_rating") val avgRating: Double? = null,
    @SerialName("is_featured") val isFeatured: Boolean? = null,
    @SerialName("created_at") val createdAt: String? = null,
    @SerialName("updated_at") val updatedAt: String? = null,
)

/** A chapter within an audiobook. */
@Serializable
data class AudiobookChapter(
    val id: String? = null,
    val title: String? = null,
    @SerialName("start_time") val startTime: Double? = null,
    @SerialName("end_time") val endTime: Double? = null,
) {
    val stableId: String get() = id ?: "${startTime ?: 0.0}"
}

/** Paginated response from GET /api/v1/audiobooks */
@Serializable
data class AudiobookListResponse(
    val items: List<Audiobook>? = null,
    val total: Int? = null,
    val page: Int? = null,
    @SerialName("page_size") val pageSize: Int? = null,
    @SerialName("total_pages") val totalPages: Int? = null,
)
